README
 WHAT IS THIS?
===============

This is a library for the Arduino for controlling LCD modules based on the
popular Hitachi HD44780 controller.  The library supports all features of
the controller and can operate in either 4pin or 8pin mode without modifying
any source code.


 WHY, THERE IS ALREADY ANOTHER LIB?
====================================

Well I thought it would be fun to write, and the other libraries where not
really fully featured and or a bit messy.  I also wanted to have a single
library for handing both 4pin and 8pin mode without modifying library files.


 HOW TO INSTALL
================

Make a directory in the arduino libraries directory:
mkdir ~/arduino/hardware/libraries/Lcd

Copy Lcd.h and Lcd.cpp to it:
cp Lcd.h Lcd.cpp ~/arduino/hardware/libraries/Lcd


 HOW TO USE
============

Click Sketch->Import Library->Lcd.  This should include the Lcd.h header
automagically.  You will find some sample sketches that demonstrate how
to use this library in the package you downloaded.


 WIRING
========

By default the Lcd class will setup the LCD module in 8 pin mode with the
following pin assignment:

RS -> 1
RW -> 2
E  -> 3
D0 -> 4
...
D7 -> 11

As of version 0.2, the RW pin is used in order to check the busy flag of
the LCD's internal processor.  This eliminates the hard coded waits after
each command sent to the LCD and should prove to be more portable and
reliable in the long run.



ChangeLog
2009-03-27 Josh Kropf <josh@slashdev.ca>

	* Lcd.h, Lcd.cpp
	  (print): added method overrides for printing numbers


2008-09-17 Josh Kropf <josh@slashdev.ca>

	* Lcd.h: use protected member to store 4bit flag instead of line function
	* Lcd.cpp: use flag to check if 4bit mode is enabled
	  (check_bf): fixed bug resulting in wrong pin selected for BF check
	  (check_bf): do second enable pin pulse in 4bit mode


2008-09-14 Josh Kropf <josh@slashdev.ca>

	* Lcd.h (send): added register to parameter list
	  (set_enable, pulse_enable): new functions
	* Lcd.cpp (send): set register after busy flag check
	  (send): move busy flag check to start of function
	  (check_bf): various fixes, but still doesn't work
	  (setup): added fixed delay in 4bit init instead of busy flag check


2008-03-09 Josh Kropf <josh@slashdev.ca>

	* Lcd.h (CTRLPINS): macro for generating controll pin mask
	* Lcd.cpp (send_8bit): new method for sending commands in 8bit mode
	(send): removed delay parameter, check BF instead
	(check_bf): new method for checking busy flag
	(enable): removed first write of LOW to E pin: E is normally LOW
	(setup): cleaned up initialization to match hd44780 spec exactly

2008-01-12 Josh Kropf <josh@slashdev.ca>

	* first release version 0.1



Lcd.cpp
/*
 * Copyright 2008 Josh Kropf
 * 
 * This file is part of the Arduino LCD Library.
 * 
 * Arduino LCD Library is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 * 
 * Arduino LCD Library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with Lemon Launcher; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 */
#include "Lcd.h"

extern "C" {
   #include <inttypes.h>
   #include "WConstants.h"
   //#include "WProgram.h"
}

/* extract 4 bit value from the mask at the given offset */
#define PIN(mask,off) (mask>>off)&0xf

/* List of available commands */
#define CMD_CLEAR      0x01
#define CMD_HOME       0x02
#define CMD_ENTRY_MODE 0x04
#define CMD_DISPLAY    0x08
#define CMD_SHIFT      0x10
#define CMD_FUNCTION   0x20
#define CMD_CGRAM_ADR  0x40
#define CMD_DGRAM_ADR  0x80

#define REG_INST 0x00 // write instruction
#define REG_DATA 0x10 // write data into ram
#define REG_READ 0x01 // read busy flag/address counter

void
Lcd::select_reg(uint8_t reg)
{
   digitalWrite(PIN(_ctrl_pins, 8), (reg&0xf0)>>4);
   digitalWrite(PIN(_ctrl_pins, 4),  reg&0x0f);
}

void
Lcd::send(uint8_t reg, uint8_t data)
{
   // when RW bin is not connected, comment out busy flag check
   // and use fixed delay (might need to adjust delay length)
   // delay(5);
   check_bf();
   
   select_reg(reg);
   
   if (_is4bit) {
      // in 4 pin mode, send upper 4 bits then lower 4 bits
      send_4bits(data >> 4);
      send_4bits(data & 0xf);
   } else {
      send_8bits(data);
   }
}

void
Lcd::send_4bits(uint8_t data)
{
   // 4 pin mode we use only the upper 16 bits (4bits x 4 pins)
   uint32_t pins = _data_pins;
   
   for (int i=0; i<4; i++) {
      digitalWrite(pins & 0xf, data & 1);
      pins >>= 4;
      data >>= 1;
   }
   
   pulse_enable();
}

void
Lcd::send_8bits(uint8_t data)
{
   uint32_t pins = _data_pins;
   
   for (int i=0; i<8; i++) {
      digitalWrite(pins & 0xf, data & 1);
      pins >>= 4;
      data >>= 1;
   }
   
   pulse_enable();
}

void
Lcd::set_enable(uint8_t val)
{
   digitalWrite(PIN(_ctrl_pins,0), val);
   delayMicroseconds(1);
}

void
Lcd::pulse_enable()
{
   digitalWrite(PIN(_ctrl_pins,0), HIGH);
   delayMicroseconds(1);
   digitalWrite(PIN(_ctrl_pins,0), LOW);
   delayMicroseconds(1);
}

void
Lcd::check_bf()
{
   // busy flag (D7 pin) is at offset 12 for 4bit and 28 for 8bit
   uint8_t bf_pin = PIN(_data_pins, (_is4bit? 12 : 28));
   
   pinMode(bf_pin, INPUT); // put D7 into input mode
   select_reg(REG_READ);   // select register for reading busy flag

   uint8_t bf;
   do {
      set_enable(HIGH);
      bf = digitalRead(bf_pin);
      set_enable(LOW);
      
      // this operation has two steps in 4bit bus mode
      // 1) get the busy flag bit and 3 bits of address counter
      // 2) remaining 4 bits of address counter
      // so do one more pulse of the enable pin to finish the op
      if (_is4bit)
         pulse_enable();
   } while (bf);
   
   pinMode(bf_pin, OUTPUT); // return D7 to output mode
}

Lcd::Lcd(uint8_t width, uint8_t func) : _cols(width), _function(func)
{
   _is4bit = !(_function & FUNCTION_8BIT);
   
   // set default pin masks
   _ctrl_pins = CTRLPINS(1,2,3);
   _data_pins = _8PINS(4,5,6,7,8,9,10,11);
}

void
Lcd::setup()
{
   uint16_t cpins = _ctrl_pins;
   uint32_t dpins = _data_pins;
   
   // setup control pins
   for (int i=0; i<3; i++) {
      pinMode(cpins&0xf, OUTPUT);
      cpins >>=4;
   }
   
   // setup data pins
   for (int i=0; i<(_is4bit? 4 : 8); i++) {
      pinMode(dpins&0xf, OUTPUT);
      dpins >>=4;
   }
   
   // spec says wait for the internal reset circuit to finish
   // before trying to do initialization by instruction
   delay(50);
   
   select_reg(REG_INST);  // use instruction register
   
   if (_is4bit) {
      send_4bits(0x03);  // function set 8bit wait 5 millis
      delay(5);

      send_4bits(0x03);  // function set 8bit wait 100 micros
      delayMicroseconds(150);

      send_4bits(0x03);  // final 8bit function set wait 100 micros
      delayMicroseconds(150);
      
      send_4bits(0x02);  // first 4bit function command, now bf can be checked
   } else {
      send_8bits(0x30);
      delay(5);

      send_8bits(0x30);
      delayMicroseconds(150);

      send_8bits(0x30);
   }
   
   // send final funtion command, can't be set again after this
   send(REG_INST, CMD_FUNCTION | _function);
   
   // remaining three instructions of the initialization routine
   // hd44780 spec states the second cmd should be display clear
   // but display on is the only one that works
   display(DISPLAY_OFF);
   display(DISPLAY_ON);
   entry_mode(ENTRY_CURSOR_INC | ENTRY_NOSHIFT);
}

void
Lcd::entry_mode(uint8_t entry)
{
   send(REG_INST, CMD_ENTRY_MODE | entry);
}

void
Lcd::display(uint8_t display)
{
   send(REG_INST, CMD_DISPLAY | display);
}

void
Lcd::shift(uint8_t shift)
{
   send(REG_INST, CMD_SHIFT | shift);
}

void
Lcd::clear()
{
   send(REG_INST, CMD_CLEAR);
}

void
Lcd::home()
{
   send(REG_INST, CMD_HOME);
}

void
Lcd::move_to(int col, int row)
{
   if (_function & FUNCTION_2LINE) {
      switch(row) {
         case 1:
            send(REG_INST, CMD_DGRAM_ADR | col-1); break;
         case 3:
            send(REG_INST, CMD_DGRAM_ADR | _cols+col-1); break;
         case 2:
            send(REG_INST, CMD_DGRAM_ADR | 0x40+col-1); break;
         case 4:
            send(REG_INST, CMD_DGRAM_ADR | 0x40+_cols+col-1 ); break;
      }
   } else {
      send(REG_INST, CMD_DGRAM_ADR | ((row-1)*_cols)+col-1);
   }
}

void
Lcd::print(char c)
{
   send(REG_DATA, c);
}

void
Lcd::print(char* s)
{
   while(*s != 0x00) {
      print(*s);
      s++;
   }
}

void
Lcd::print(unsigned long n, uint8_t base)
{
   char buf[8 * sizeof(long)];
   int i = 0;

   if (n == 0) {
      print('0');
      return;
   }

   while (n > 0) {
      buf[i++] = n % base;
      n /= base;
   }

   for(; i > 0; i--) {
      print((char) (buf[i - 1] < 10?
               '0' + buf[i - 1] :
               'A' + buf[i - 1] - 10));
   }
}

void
Lcd::print(long n, uint8_t base)
{
   if (n < 0) {
      print('-');
      n = -n;
   }

   print((unsigned long)n);
}

void
Lcd::define_char(uint8_t index, uint8_t data[])
{
   int h = _function & FUNCTION_5x11? 11 : 8;
   for (int i=0; i<h; i++) {
      send(REG_INST, CMD_CGRAM_ADR | (index*h)+i);
      send(REG_DATA, data[i]);
   }
}


Lcd.h
/*
 * Copyright 2008 Josh Kropf
 * 
 * This file is part of the Arduino LCD Library.
 * 
 * Arduino LCD Library is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 * 
 * Arduino LCD Library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with Lemon Launcher; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 */
#ifndef _LCD_H_
#define _LCD_H_

#include <inttypes.h>

/* Generates control pin mask */
#define CTRLPINS(RS,RW,E) \
   ((RS&0xf) << 8) | ((RW&0xf) << 4) | (E&0xf)

/* Generates data pin mask for 4 pin mode */
#define _4PINS(D4,D5,D6,D7) \
   ((D7&0xf) << 12) | ((D6&0xf) << 8) | \
   ((D5&0xf) << 4)  |  (D4&0xf)

/* Generates data pin mask for 8 pin mode */
#define _8PINS(D0,D1,D2,D3,D4,D5,D6,D7) \
   (uint32_t)((((uint32_t)D7)&0xf) << 28) | (uint32_t)((((uint32_t)D6)&0xf) << 24) | \
   (uint32_t)((((uint32_t)D5)&0xf) << 20) | (uint32_t)((((uint32_t)D4)&0xf) << 16) | \
   (uint32_t)((((uint32_t)D3)&0xf) << 12) | (uint32_t)((((uint32_t)D2)&0xf) << 8) | \
   (uint32_t)((((uint32_t)D1)&0xf) << 4)  | (uint32_t)(D0&0xf)

// Function constants passed to constructor, can be OR'd together
#define FUNCTION_8BIT  0x10 // enable 8 pin mode
#define FUNCTION_4BIT  0x00 // enable 4 pin mode
#define FUNCTION_2LINE 0x08 // LCD has two lines, line two starts at addr 0x40
#define FUNCTION_1LINE 0x00 // LCD has one continues line
#define FUNCTION_5x11  0x04 // use 5x11 custom characters
#define FUNCTION_5x8   0x00 // use 5x8 custom characters

// CMD_DISPLAY parameters, can be OR'd together
#define DISPLAY_ON  0x04      // display on
#define DISPLAY_OFF 0x00      // display off
#define DISPLAY_CURSOR   0x02 // cursor on
#define DISPLAY_NOCURSOR 0x00 // cursor off
#define DISPLAY_BLINK   0x01  // cursor blink on
#define DISPLAY_NOBLINK 0x00  // curson blink off

// CMD_ENTRY_MODE parameters, can be OR'd together
#define ENTRY_CURSOR_INC 0x02 // increment cursor position
#define ENTRY_CURSOR_DEC 0x00 // decrement curson position
#define ENTRY_SHIFT      0x01 // shift entire display
#define ENTRY_NOSHIFT    0x00 // don't shift display

// CMD_SHIFT parameters, can be OR'd together
#define SHIFT_SCREEN 0x08 // shift display 
#define SHIFT_CURSOR 0x00 // shift cursor
#define SHIFT_RIGHT  0x04 // to the right
#define SHIFT_LEFT   0x00 // to the left

/**
 * Class for controlling HD44780 based LCD modules from the Arduino.
 *
 * The Arduino has 14 pins.  Therefore we can easily encode a single pin
 * number inside 4bits (0-15).  This class uses a 16 bit mask to encode
 * the control pins (RS,RW,E) and a 32 bit mask to encode the data
 * pins (D0~D8).
 *
 * Control Pins
 * Nibble  | N2 | N1 | N0 |
 * --------+----+----+----+
 * LCD Pin | RS | RW |  E |
 *
 * Default pin arrangement is:
 * RS=1,RW=2,Enable=3
 *
 * Data Pins
 * Nibble  | N7 | N6 | ... | N0 |
 * --------+----+----+-----+----+
 * LCD Pin | D0 | D1 | ... | D7 |
 *
 * Default pin arrangement is (4pin mode uses D4~D7):
 * D0=4, D1=5... , D7=11
 *
 * Version: 0.4
 */
class Lcd {
protected:
   bool _is4bit;   // flag indicating 4bit mode is enabled
   
   uint8_t  _cols; // number of columns wide
   uint8_t  _function;
   
   uint16_t _ctrl_pins;
   uint32_t _data_pins;
   
   /*
    * Select register and send 8bits to the LCD.  Before the
    * operation begins, the busy flag is checked and does not
    * proceed until BF pin (D7) is low.
    */
   void send(uint8_t reg, uint8_t data);
   
   /* sends the lower 4bits to the lcd on pins D7~D4 */
   void send_4bits(uint8_t data);

   /* sends all 8bits to the lcd on pins D7~D0 */
   void send_8bits(uint8_t data);
   
   /* select register (REG_XXXX constant) */
   void select_reg(uint8_t reg);

   /* sets the enable pin to HIGH or LOW and adds a short delay */
   inline void set_enable(uint8_t val);
   
   /* toggle the enable pin HIGH then LOW */
   inline void pulse_enable();

   /* wait for BF to fall LOW */
   void check_bf();
   
public:
   /**
    * Creates the Lcd class.
    * \param width number of columns in the LCD module
    * \param func function parameters (mask of FUNCTION_* defines)
    */
   Lcd(uint8_t width, uint8_t func = 0);
   
   /**
    * Setup Arduino pins and initialize LCD module.
    */
   void setup();
   
   /**
    * Sets the 8 bit mask for the control pins (RS,RW,E).
    */
   inline void set_ctrl_pins(uint16_t mask)
   { _ctrl_pins = mask; }

   /**
    * Sets the 32 bit mask for the data pins.  When using 4 wire mode, only use
    * the upper 4 pins (D4~D7) in the lower 4 bytes (B3~B0).
    */
   inline void set_data_pins(uint32_t mask)
   { _data_pins = mask; }

   /**
    * Sets entry mode parameters.  Combination of ENTRY_* defines.
    */
   void entry_mode(uint8_t entry);
   
   /**
    * Sets display mode parameters.  Combination of DISPAY_* defines.
    */
   void display(uint8_t display);

   /**
    * Sets cursor or display shift dehaviour.  Combination of SHIFT_* defines.
    */
   void shift(uint8_t shift);

   /**
    * Clear the display.
    */
   void clear();

   /**
    * Move cursor to home position (top/left).
    */
   void home();
   
   /**
    * Move the cursor to a given column and row.
    * \param col
    * \param row
    */
   void move_to(int col, int row);

   /**
    * Print a single ASCII character to the LCD screen at the current
    * cursor position.
    */
   void print(char c);

   /**
    * Print a string of ASCII characters to the LCD screen at the
    * current cursor position.
    */
   void print(char* str);

   /**
    * Print a number to the LCD screen at the current cursor position.
    * \param n unsigned long integer
    * \param base number system, default is decimal (base ten)
    */
   void print(unsigned long n, uint8_t base = 10);

   /**
    * Print a number to the LCD screen at the current cursor position.
    * \param n signed long integer
    * \param base number system, default is decimal (base ten)
    */
   void print(long n, uint8_t base = 10);

   /**
    * Print a number to the LCD screen at the current cursor position.
    * \param n unsigned integer
    * \param base number system, default is decimal (base ten)
    */
   void print(unsigned int n, uint8_t base = 10)
   { print((unsigned long)n, base); }

   /**
    * Print a number to the LCD screen at the current cursor position.
    * \param n signed integer
    * \param base number system, default is decimal (base ten)
    */
   void print(int n, uint8_t base = 10)
   { print((long)n, base); }

   /**
    * Define character in CGRAM.  This defines a custom character.  When the
    * LCD is setup to use 5x8 font the user can specify up to eight characters.
    * For 5x11 font four characters can be defined.
    *
    * Custom characters are bitmaps!  For 5x8 font the bitmap must be 8 bits
    * wide and 8 bits hight.  For 5x11 font the bitmap must be 8 bits wide and
    * 10 bits high.
    * font the array must contain 11 bytes.
    *
    * Each byte in the character data array represents a row of bits in a
    * character.  The lower 5 bits of each byte is used to specify the dots in
    * the character that should be black (1) and white (0).
    *
    * To print the user defined character, pass the index of of the character
    * to the print method.
    *
    * Example: define bullet character
    *
    * X X X 0 0 0 0 0
    * X X X 0 0 0 0 0
    * X X X 0 1 1 1 0
    * X X X 1 1 1 1 1
    * X X X 1 1 1 1 1
    * X X X 0 1 1 1 0
    * X X X 0 0 0 0 0
    * X X X 0 0 0 0 0
    * 
    * {0x00, 0x00, 0x0E, 0x1F, 0x1F, 0x0E, 0x00, 0x00}
    *
    * \param index zero based index of custom character
    * \param data character data array
    */
   void define_char(uint8_t index, uint8_t data[]);
};

#endif
